pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -1531,3 +1532,80 @@ def _handle_distribution_tree(declarative_content):
+# The most reasonable means of fixing that requires adjusting existing content units
+class ModifiedQueryExistingContents(Stage):
+    """
+    A Stages API stage that saves :attr:`DeclarativeContent.content` objects and saves its related
+    :class:`~pulpcore.plugin.models.ContentArtifact` objects too.
+
+    This stage expects :class:`~pulpcore.plugin.stages.DeclarativeContent` units from `self._in_q`
+    and inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    This stage inspects any "unsaved" Content unit objects and searches for existing saved Content
+    units inside Pulp with the same unit key. Any existing Content objects found, replace their
+    "unsaved" counterpart in the :class:`~pulpcore.plugin.stages.DeclarativeContent` object.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `self._out_q` after it has
+    been handled.
+
+    This stage drains all available items from `self._in_q` and batches everything into one large
+    call to the db for efficiency.
+    """
+
+    async def run(self):
+        """
+        The coroutine for this stage.
+
+        Returns:
+            The coroutine for this stage.
+        """
+        async for batch in self.batches():
+            content_q_by_type = defaultdict(lambda: Q(pk__in=[]))
+            d_content_by_nat_key = defaultdict(list)
+            for d_content in batch:
+                if d_content.content._state.adding:
+                    model_type = type(d_content.content)
+                    unit_q = d_content.content.q()
+                    content_q_by_type[model_type] = content_q_by_type[model_type] | unit_q
+                    d_content_by_nat_key[d_content.content.natural_key()].append(d_content)
+
+            for model_type, content_q in content_q_by_type.items():
+                try:
+                    await sync_to_async(model_type.objects.filter(content_q).touch)()
+                except AttributeError:
+                    raise TypeError(
+                        "Plugins which declare custom ORM managers on their content classes "
+                        "should have those managers inherit from "
+                        "pulpcore.plugin.models.ContentManager."
+                    )
+                async for result in sync_to_async_iterable(
+                    model_type.objects.filter(content_q).iterator()
+                ):
+                    for d_content in d_content_by_nat_key[result.natural_key()]:
+                        # ============ The below lines are added ============
+                        if model_type == Package:
+                            incorrect_changelogs = result.changelogs != d_content.content.changelogs
+                            incorrect_modular_relation = (
```

Comment:
this object `d_content.content` already has trimmed changelogs, right?

Reply:
Right, it's an incoming `Package` object which has been created, and all of that work should have already occurred.